spec/dds_md5_reporter_spec.rb

```diff
@@ -164,42 +227,291 @@
+        end
+      end
+
+      describe 'behavior' do
+        context 'with incorrect credentials' do
+          include_context 'auth_token request with error'
 
           it {
-            expect(subject.auth_token).to eq(expected_api_token)
+            expect {
+              subject.auth_token
+            }.to raise_error(expected_error)
           }
         end
 
-        context 'when token is expired' do
-          let(:expected_calls) { 2 }
-          before do
-            subject.auth_token
+        context 'with correct credentials' do
+          include_context 'successfull auth_token request'
+
+          context 'for the first time' do
+            it 'should return an auth_token' do
+              expect(subject.auth_token).to eq(expected_api_token)
+            end
           end
 
-          it {
-            travel_to(Time.now + 10000) do
+          context 'when token is not expired' do
+            before do
+              subject.auth_token
+            end
+
+            it {
               expect(subject.auth_token).to eq(expected_api_token)
+            }
+          end
+
+          context 'when token is expired' do
+            let(:expected_calls) { 2 }
+            before do
+              subject.auth_token
             end
+
+            it {
+              travel_to(Time.now + 10000) do
+                expect(subject.auth_token).to eq(expected_api_token)
+              end
+            }
+          end
+        end
+      end
+    end
+
+    describe '#auth_header' do
+      it { is_expected.to respond_to(:auth_header) }
+
+      describe 'behavior' do
+        before do
+          is_expected.to receive(:auth_token)
+            .and_return(expected_api_token)
+          is_expected.to receive(:json_headers)
+            .and_return(expected_json_headers)
+        end
+        it {
+          expect(subject.auth_header).to eq(expected_auth_header)
+        }
+      end
+    end
+
+    shared_context 'a dds token authenticated request' do
+      let(:expected_request_headers) {
+        expected_auth_header
+      }
+
+      before do
+        expect(reporter).to receive(:auth_header)
+          .and_return(expected_auth_header)
+      end
+    end
+
+    shared_examples 'a failed dds api request' do
+      let(:failure_code) { "400" }
+      let(:expected_error) { StandardError.new("failed dds api request") }
+      let(:expected_response) {
+        instance_double("HTTParty::Response")
+      }
+      let(:expected_response_response) {
+        double()
+      }
+
+      before do
+        expect(expected_response).to receive(:response)
+          .and_return(expected_response_response)
+        expect(expected_response_response).to receive(:code)
+          .and_return(failure_code)
+        expect(reporter).to receive(:raise_dds_api_exception)
+          .with(expected_preamble, expected_response)
+          .and_raise(expected_error)
+        if (expected_http_verb == :get)
```

Comment:
It's best to avoid branching logic in tests.

Reply:
ok. I removed this already, as I refactored this another way.